libgalois/src/analytics/k_truss/k_truss.cpp

```diff
@@ -0,0 +1,459 @@
+    size_t curSize = g->num_nodes(), nextSize;
+
+    katana::do_all(
+        katana::iterate(*g), KeepValidNodes{g, k, *next}, katana::steal());
+    nextSize = std::distance(next->begin(), next->end());
+
+    while (curSize != nextSize) {
+      cur->clear();
+      curSize = nextSize;
+      std::swap(cur, next);
+
+      katana::do_all(
+          katana::iterate(*cur), KeepValidNodes{g, k, *next}, katana::steal());
+      nextSize = std::distance(next->begin(), next->end());
+    }
+  }
+};  ///< End BSPCoreAlgo.
+
+/**
+ * BSPCoreThenTrussAlgo:
+ * 1. Reduce the graph to k-1 core
+ * 2. Compute k-truss from k-1 core
+ */
+struct BSPCoreThenTrussAlgo {
+  std::string name() { return "bspCoreThenTruss"; }
+
+  void operator()(Graph* g, unsigned int k) {
+    if (k - 2 == 0) {
+      return;
+    }
+
+    katana::StatTimer TCore("Reduce_to_(k-1)-core");
+    TCore.start();
+
+    BSPCoreAlgo bspCore;
+    bspCore(g, k - 1);
+
+    TCore.stop();
+
+    katana::StatTimer TTruss("Reduce_to_k-truss");
+    TTruss.start();
+
+    BSPTrussAlgo bspTrussIm;
+    bspTrussIm(g, k);
+
+    TTruss.stop();
+  }  ///< End operator().
+};   ///< End struct BSPCoreThenTrussAlgo.
+
+template <typename Algorithm>
+static katana::Result<void>
+KTrussWithWrap(
+    katana::PropertyFileGraph* pfg, std::string output_property_name,
+    uint32_t k_truss_number) {
+  auto pg_result = Graph::Make(pfg, {}, {output_property_name});
+  if (!pg_result) {
+    return pg_result.error();
+  }
+  auto graph = pg_result.value();
+
+  KTrussInitialization(&graph);
+  Algorithm algo;
+
+  katana::StatTimer exec_time("KTruss");
+  exec_time.start();
+  algo(&graph, k_truss_number);
+  exec_time.stop();
+
+  return katana::ResultSuccess();
+}
+
+katana::Result<void>
+katana::analytics::KTruss(
+    katana::PropertyFileGraph* pfg, uint32_t k_truss_number,
+    const std::string& output_property_name, KTrussPlan plan) {
+  if (auto result =
+          ConstructEdgeProperties<EdgeData>(pfg, {output_property_name});
+      !result) {
+    return result.error();
+  }
+
+  auto result = katana::SortAllEdgesByDest(pfg);
+  if (!result) {
+    return result.error();
+  }
+
+  switch (plan.algorithm()) {
+  case KTrussPlan::kBsp:
+    return KTrussWithWrap<BSPTrussAlgo>(
+        pfg, output_property_name, k_truss_number);
+    break;
+  case KTrussPlan::kBspJacobi:
+    return KTrussWithWrap<BSPTrussJacobiAlgo>(
+        pfg, output_property_name, k_truss_number);
+    break;
+  case KTrussPlan::kBspCoreThenTruss:
+    return KTrussWithWrap<BSPCoreThenTrussAlgo>(
+        pfg, output_property_name, k_truss_number);
+    break;
+  default:
+    return ErrorCode::InvalidArgument;
+  }
+}
+
+// TODO (gill) Add a validity routine.
+katana::Result<void>
+katana::analytics::KTrussAssertValid(
+    [[maybe_unused]] katana::PropertyFileGraph* pfg,
+    [[maybe_unused]] uint32_t k_truss_number,
+    [[maybe_unused]] const std::string& property_name) {
+  return katana::ResultSuccess();
+}
```

Comment:
Should we include these functions in the header if they are not implemented? It feels like it might create a false sense of security, so I would prefer to omit them from the API for the moment. We would have a TODO in the header to create an `AssertValid` routine.